internal/crdReplicator/crdReplicator-operator.go

```diff
@@ -0,0 +1,704 @@
+	LocalDynClient dynamic.Interface
+	//a list of GVRs of resources to be replicated
+	RegisteredResources []schema.GroupVersionResource
+	//each time a resource is removed from the configuration it is saved in this list,
+	//it stays here until the associated watcher, if running, is stopped
+	UnregisteredResources []string
+	//for each local resource that exists a running watcher we save the tuples (registeredResource, chan)
+	LocalWatchers map[string]chan bool
+	//for each peering cluster we save all the running watchers monitoring the replicated resources:(clusterID, (registeredResource, chan))
+	RemoteWatchers map[string]map[string]chan bool
+}
+
+func (d *CRDReplicatorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var fc v1.ForeignCluster
+	ctx := context.Background()
+
+	err := d.Get(ctx, req.NamespacedName, &fc)
+	if err != nil && !apierrors.IsNotFound(err) {
+		klog.Errorf("%s -> unable to retrieve resource %s: %s", d.ClusterID, req.NamespacedName, err)
+		return result, err
+	}
+	if apierrors.IsNotFound(err) {
+		klog.Errorf("%s -> resource %s not present, probably deleted: %s", d.ClusterID, req.NamespacedName, err)
+		return ctrl.Result{}, nil
+	}
+	remoteClusterID := fc.Spec.ClusterID
+	//check if the config of the peering cluster is ready
+	//first we check the outgoing connection
+	if fc.Status.Outgoing.AvailableIdentity {
+		//retrieve the config
+		config, err := d.getConfig(d.ClientSet, fc.Status.Outgoing.IdentityRef, remoteClusterID)
+		if err != nil {
+			klog.Errorf("%s -> unable to retrieve config from resource %s for remote peering cluster %s: %s", d.ClusterID, req.NamespacedName, remoteClusterID, err)
+			return result, nil
+		}
+		//check if the dynamic client exists
+		if _, ok := d.RemoteDynClients[remoteClusterID]; !ok {
+			dynClient, err := dynamic.NewForConfig(config)
+			if err != nil {
+				klog.Errorf("%s -> unable to create dynamic client: %s", remoteClusterID, err)
+				return result, err
+			} else {
+				klog.Infof("%s -> dynamic client created", remoteClusterID)
+			}
+			d.RemoteDynClients[remoteClusterID] = dynClient
+		}
+	} else if fc.Status.Incoming.AvailableIdentity {
+		//retrieve the config
+		config, err := d.getConfig(d.ClientSet, fc.Status.Outgoing.IdentityRef, remoteClusterID)
+		if err != nil {
+			klog.Errorf("%s -> unable to retrieve config from resource %s for remote peering cluster %s: %s", d.ClusterID, req.NamespacedName, remoteClusterID, err)
+			return result, err
+		}
+		//check if the dynamic client exists
+		if _, ok := d.RemoteDynClients[remoteClusterID]; !ok {
+			dynClient, err := dynamic.NewForConfig(config)
+			if err != nil {
+				klog.Errorf("%s -> unable to create dynamic client: %s", remoteClusterID, err)
+				return result, err
+			} else {
+				klog.Infof("%s -> dynamic client created", remoteClusterID)
+			}
+			d.RemoteDynClients[remoteClusterID] = dynClient
+		}
+	}
+
+	d.StartWatchers()
+	d.StartRemoteWatchers()
+	d.StopWatchers()
```

Comment:
can `defer` on these functions be useful? (in case an error occurs we are sure the watchers are closed)

Reply:
the watchers need to be closed only if the related resource has been removed from the configuration.
if we the secret containing the kubeconfig is not ready it means that watchers for that peering cluster never started so we have nothing to close.